### ISSUE TITLE
Unnecessary spaces in UTF-8 encoding

### ISSUE THREAD

**Issue by Vladislav4KZ** [FWGS/xash3d-fwgs#344]
HUD with Windows-1251 charset
==================================================================
![Windows-1251_HUD](https://user-images.githubusercontent.com/22411953/110277009-9f8a6500-7ffe-11eb-921c-333fc5f850e3.png)

HUD with UTF-8 charset
==================================================================
![UTF-8_HUD](https://user-images.githubusercontent.com/22411953/110277032-aadd9080-7ffe-11eb-8d6c-a3f905d8c0ec.png)

VGUI with Windows-1251 charset
==================================================================
![Windows-1251_VGUI](https://user-images.githubusercontent.com/22411953/110277057-bcbf3380-7ffe-11eb-8f2e-6d0f7ae9a077.png)

VGUI with UTF-8 charset
==================================================================
![UTF-8_VGUI](https://user-images.githubusercontent.com/22411953/110277086-cf396d00-7ffe-11eb-95d3-6da6c43aba18.png)

**Comment by mittorn** [comment]
Для vgui весьма проблематично это исправить. vgui рисует текст посимвольно. Возможно потребуется класс для шрифта реализовывать как минимум

**Comment by mittorn** [comment]
Касательно hud - вероятно, баг

**Comment by a1batross** [comment]
В феврале этого года я отрефакторил весь код который печатает символы, строчки и так далее (используя движковые шрифты, типа консольного или HUD), и подобная проблема больше не должна проявляться.

Это кстати оригинальный клиент TFC?

**Comment by Vladislav4KZ** [comment]
> Это кстати оригинальный клиент TFC?

Да

> В феврале этого года я отрефакторил весь код который печатает символы, строчки и так далее (используя движковые шрифты, типа консольного или HUD), и подобная проблема больше не должна проявляться.

Проверю позже

**Comment by Vladislav4KZ** [comment]
@a1batross поставил самый последний билд ксаша, и перепроверил.

Текст, содержащий кириллицу, сохранённый в кодировке UTF-8, всё ещё отображается с пробелами между букв.

Это проявляется как во VGUI, так и в HUD.

**Comment by mittorn** [comment]
А с каких пор такое в HUD? Помню только в vgui, но чтобы это решить - придётся перериализовать несколько классов от vgui в vgui_support насколько мне известно. Сейчас от vgui приходят только компнды отрисовки отдельных символов, а не строк, потому так получается, надо зайти глубже и вклиниться чуть раньше

**Comment by Vladislav4KZ** [comment]
> А с каких пор такое в HUD?

В HUD всегда такое было.

Например, это можно заметить и в старом движке, подключившись к какому-нибудь серверу в CS16Client который отображает кириллицу в HUD

**Comment by a1batross** [comment]
@mittorn эту проблему я уже решил в freevgui (осталось его доделать :)).

В HUD похоже такие проблемы в оригинальных клиентках.

**Comment by a1batross** [comment]
The bug is definitely is in client.dll.

But I'm not sure if it's our bug or not. It's we who added UTF-8 support without providing a real way to query text size.

Yes, there are DrawString calls added by Valve but this won't help to center text or draw it right aligned, unless we're calling them outside the screen and engine skips drawing and just returns the length (which is something that Xash mods already use).

**Comment by SNMetamorph** [comment]
Догадываюсь, в чём причина бага. В корректном случае, текст в меню рисуется при помоши `DrawUtils::DrawHudString`, в таком случае всё выглядит как положено. Другой вид текста на экране рендерится ПОБАЙТОВО через `DrawUtils::TextMessageDrawChar`, понятное дело на UTF-8 строках оно не будет давать корректного результата. Всё это тестил я на cs16client свежем.

**Comment by mittorn** [comment]
Да, так и есть. В hlsdk я заменял все вызовы отрисовки на api, рисующее строки целиком когда исправлял юникод. cs16client ответвился раньше и там мог остаться старый код. Ещё в кс включен hud_utf8, это тоже может создавать проблемы.
В итоге получилась такая конструкция: если cl_charset выставлен в 1251 как и  con_chaset - имитируется поведение, на которые полагаются воновские русификаторы, если cl_charset utf-8, но hud_utf8 0 - то работает гибридный режим, который для старых api hud (которые рисуют посимвольно) оставляет 1251 чтобы не ломать старые клиенты/моды которые могли полагаться на шрифт из gfx.wad. Ну и если включить hud_utf8 то движок врубает utf8 для всех API, включая посимвольные - это по идее полезно для мультиплеера чтобы заставить работать utf-8 со всеми клиентами

**Comment by nekonomicon** [comment]
На самом деле cs16-client основан на репозитории hlsdk-client.
hlsdk-xash3d уже вроде как был начат позже почти с нуля, не считая того факта, что он на дядьМишином hlsdk, а hlsdk-client на обычном 2.3.

**Comment by mittorn** [comment]
Ну тем более там это не учтено. Подробностей не помню уже
